carsus/io/cmfgen/base.py

```diff
@@ -0,0 +1,440 @@
+        String to search.
+    string2 : str
+        Extra string to search (default is '').
+    how : {'one', 'both', 'first'}
+        If 'both' search for string1 AND string2. If 'one' search for string1\
+            OR string2. If 'first' searches for 'string1' AND NOT string2\
+            (default is 'both').
+    num_row : bool
+        If true, returns row number instead (default is False).
+
+    Returns
+    -------
+    str or int
+        Returns matching line or row number.
+
+    """
+    with open(fname, encoding='ISO-8859-1') as f:
+        n = 0
+        for line in f:
+            n += 1
+
+            if how == 'one':
+                if string1 in line or string2 in line:
+                    break
+
+            if how == 'both':
+                if string1 in line and string2 in line:
+                    break
+
+            if how == 'first':
+                if string1 in line and string2 not in line:
+                    break
+
+        # In case there's no match
+        else:
+            n = None
+            line = None
+
+    if num_row is True:
+        return n
+
+    return line
+
+
+def parse_header(fname, keys, start=0, stop=50):
+    """Parse header from CMFGEN files.
+
+    Parameters
+    ----------
+    fname : str
+        Path to plain text file.
+    keys : list of str
+        Entries to search.
+    start : int
+        First line to search in (default is 0).
+    stop : int
+        Last line to search in (default is 50).
+
+    Returns
+    -------
+    dict
+        Dictionary containing metadata.
+
+    """
+    meta = {k.strip('!'): None for k in keys}
+    with gzip.open(fname, 'rt') if fname.endswith('.gz') else open(fname, encoding='ISO-8859-1') as f:
+        for line in itertools.islice(f, start, stop):
+            for k in keys:
+                if k.lower() in line.lower():
+                    meta[k.strip('!')] = line.split()[0]
+
+    return meta
+
+
+def to_float(string):
+    """ String to float, taking care of Fortran 'D' values
+
+    Parameters
+    ----------
+    string : str
+
+    """
+    try:
+        value = float(string.replace('D', 'E'))
+
+    except ValueError:
+
+        if string == '1-.00':      # Bad value at MG/VIII/23oct02/phot_sm_3000 line 23340
+            value = 10.00
+
+        if string == '*********':  # Bad values at SUL/V/08jul99/phot_op.big lines 9255-9257
+            value = np.nan
+
+    return value
+
+
+class CMFGENEnergyLevelsParser(BaseParser):
+    """
+        Description
+        ----------
+        base : pandas.DataFrame
+        columns : list of str
+            (default value = ['Configuration', 'g', 'E(cm^-1)', 'eV', 'Hz 10^15', 'Lam(A)'])
+        meta : dict
+            Metadata parsed from file header.
+
+        Methods
+        -------
+        load(fname)
+            Parses the input data and stores the results in the `base` attribute.
+    """
+
+    keys = ['!Date',  # Metadata to parse from header. TODO: look for more keys
```

Comment:
you don't really need the `!` but they do not hurt.

Reply:
Yes, I know. I wasn't sure if simple words like 'Date' can be located twice at some files with long descriptions. The '!' ensures we are looking for the header values and not some lost word in the text. For complicated strings like 'Screened nuclear charge' the exclamation mark shouldn't be necessary.